### PR TITLE
many: add new (hidden) `snap debug ensure-state-soon` command and use in tests

### PR DESCRIPTION
**Pull request by mvo5**
This makes tests more robust by allowing to trigger the ensure()
loop without creating a change or starting/stopping the daemon.

Slightly undertested currently, sorry for that, can fix that.

### CODE REVIEW

File: tests/main/classic-ubuntu-core-transition-auth/task.yaml [hunk 1]
```diff
@@ -18,9 +18,8 @@ execute: |
     mkdir -p /home/test/.snap/
     echo '{}' > /home/test/.snap/auth.json
     
-    echo "Start/stop snapd so that the transition is triggered"
-    systemctl stop snapd.service snapd.socket
-    systemctl start snapd.service snapd.socket
+    echo "Ensure transition is triggered"
+    snap ensure-state-soon
```
Comment (pedronis) [inline tests/main/classic-ubuntu-core-transition-auth/task.yaml]: is this right now? or it needs debug in front?

File: client/client.go [hunk 2]
```diff
@@ -516,3 +516,20 @@ func (client *Client) Users() ([]*User, error) {
...
 }
+
+type debugAction struct {
+	Action string `json:"action"`
+}
+
+// EnsureStateSoon is only useful when writing test code, it will trigger
+// the internal "ensure" loop of the state engine to be run as soon as
+// possible.
+func (client *Client) EnsureStateSoon() error {
```
Comment (niemeyer) [inline client/client.go]: Can we please make this into:

`func (client *Client) Debug(action string, params interface{}, result interface{}) error`

So we're not cooking a static API assuming internal implementation details that will break applications when it gets removed?

File: tests/main/classic-ubuntu-core-transition-auth/task.yaml [hunk 3]
```diff
@@ -18,11 +18,8 @@ execute: |
     mkdir -p /home/test/.snap/
     echo '{}' > /home/test/.snap/auth.json
     
-    echo "Start/stop snapd so that the transition is triggered"
-    if ! snap changes | grep "Transition ubuntu-core to core"; then
-        systemctl stop snapd.service snapd.socket
-        systemctl start snapd.service snapd.socket
-    fi
+    echo "Ensure transition is triggered"
+    snap debug ensure-state-soon
```
Comment (zyga) [inline tests/main/classic-ubuntu-core-transition-auth/task.yaml]: Nice :-)

File: cmd/snap/cmd_ensure_state_soon.go [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+		})
+	cmd.hidden = true
+}
+
+func (x *cmdEnsureStateSoon) Execute(args []string) error {
+	if len(args) > 0 {
+		return ErrExtraArgs
+	}
+
+	return Client().EnsureStateSoon()
```
Comment (niemeyer) [inline cmd/snap/cmd_ensure_state_soon.go]: This would then be:
```
return Client().Debug("ensure-state-soon", nil, nil)

File: daemon/api.go [hunk 5]
```diff
@@ -175,6 +176,11 @@ var (
 		GET:    getChanges,
 	}
 
+	debugCmd = &Command{
+		Path: "/v2/debug",
```
Comment (niemeyer) [inline daemon/api.go]: That's nice as an API.

File: daemon/api_test.go [hunk 6]
```diff
@@ -5024,3 +5024,24 @@ func (s *apiSuite) TestAliases(c *check.C) {
...
 
 }
+
+var _ = check.Suite(&postDebugSuite{})
+
+type postDebugSuite struct {
+	apiBaseSuite
+}
+
+func (s *postDebugSuite) TestPostDebugEnsureStateSoon(c *check.C) {
```
Comment (niemeyer) [inline daemon/api_test.go]: This is not actually checking that the intended action was done, I think? Should be trivial to do that given the existing mocking around ensureStateSoon.

File: client/client.go (hunk 2 above)
Comment (mvo5) [inline client/client.go]: Thanks, this makes a lot of sense. I changed the code accordingly now.

File: cmd/snap/cmd_ensure_state_soon.go (hunk 4 above)
Comment (mvo5) [inline cmd/snap/cmd_ensure_state_soon.go]: Updated.

File: daemon/api_test.go (hunk 6 above)
Comment (mvo5) [inline daemon/api_test.go]: Indeed, I updated the test to check that the action was actually triggered.

**Pull request merged**